Update rfc2136 provider to split out changes per zone

gregsidelinger opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**
Update the rfc2136 provider to update each zone separately as the current implantation causes errors if the change spans zones.

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #4106 

**Checklist**

- [ ] Unit tests updated
- [ ] End user documentation updated

mloiseleur commented:
This PR needs a test case, to reproduce this issue and ensure the fix will continue to work.

gregsidelinger commented:
@mloiseleur I have some ideas how to test for this.  It will be a few days before I get around to mocking something out but I will update the test cases.

gregsidelinger commented:
@mloiseleur I added some different test cases.  Most of the test cases I added was around things that here not being tested before.  I'm actually testing the test version of SendMessage to make sure it does not receive a multizone change, except for the root zone as I needed to update my change to address that.

mloiseleur commented:
It's looking good, thanks. I'll wait that you have finished for a complete review.
Would you please also update PR's description (first line) ? :)
/ok-to-test

gregsidelinger commented:
I think I figured out why the test suite is failing.  The order of the createMsgs and other objects updated by the stubbed version of SendMessage is not fixed as I loop over map which is not an ordered object.  Will get the suite updated shortly to address.

gregsidelinger commented:
@mloiseleur how is the PR looking now that I extended the test suite?

mloiseleur commented on provider/rfc2136/rfc2136.go:
```diff
@@ -265,26 +270,33 @@ func (r rfc2136Provider) ApplyChanges(ctx context.Context, changes *plan.Changes
+			if len(z.Ns) > 0 {
+				err := r.actions.SendMessage(z)
+				if err != nil {
+					log.Errorf("RFC2136 update failed: %v", err)
```
```suggestion
					log.Errorf("RFC2136 create failed: %v", err)
```

mloiseleur commented on provider/rfc2136/rfc2136.go:
```diff
@@ -323,18 +341,20 @@ func (r rfc2136Provider) ApplyChanges(ctx context.Context, changes *plan.Changes
+			if len(z.Ns) > 0 {
+				err := r.actions.SendMessage(z)
+				if err != nil {
+					log.Errorf("RFC2136 update failed: %v", err)
```
```suggestion
					log.Errorf("RFC2136 delete record failed: %v", err)
```

mloiseleur commented on provider/rfc2136/rfc2136.go:
```diff
@@ -265,26 +270,33 @@ func (r rfc2136Provider) ApplyChanges(ctx context.Context, changes *plan.Changes
-				continue
+		for _, z := range m {
+			if len(z.Ns) > 0 {
+				err := r.actions.SendMessage(z)
+				if err != nil {
```
```suggestion
				if err := r.actions.SendMessage(z); err != nil {
```

mloiseleur replied:
```suggestion
				if err := r.actions.SendMessage(z); err != nil {
```

mloiseleur commented on provider/rfc2136/rfc2136.go:
```diff
@@ -294,27 +306,33 @@ func (r rfc2136Provider) ApplyChanges(ctx context.Context, changes *plan.Changes
-				continue
+		for _, z := range m {
+			if len(z.Ns) > 0 {
+				err := r.actions.SendMessage(z)
+				if err != nil {
```
```suggestion
				if err := r.actions.SendMessage(z); err != nil {
```

mloiseleur commented on provider/rfc2136/rfc2136.go:
```diff
@@ -323,18 +341,20 @@ func (r rfc2136Provider) ApplyChanges(ctx context.Context, changes *plan.Changes
-				continue
+		for _, z := range m {
+			if len(z.Ns) > 0 {
+				err := r.actions.SendMessage(z)
+				if err != nil {
```
```suggestion
				if err := r.actions.SendMessage(z); err != nil {
```

mloiseleur commented:
It looks good to me. I have just minor suggestion.

gregsidelinger commented:
Ok, suggestions are in there as well as one other log msg update to make them all consistent.  And the linter check finally passed after getting kicked.

mloiseleur commented:
/lgtm

szuecs commented:
/approve
